Added public APIs for deleting blobs.

vcolin7 opened the pull request:
Addresses #195.

anuchandy commented on sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobServiceImpl.java:
```diff
@@ -1335,6 +1446,96 @@ public void onFailure(Call<ResponseBody> call, Throwable t) {
+            Response<Void> response = executeCall(call);
+
+            if (response.isSuccessful()) {
+                if (response.code() == 200) {
```
The java SDK API is [here](https://github.com/Azure/azure-sdk-for-java/blob/4871ff56eec2e0615973a6a32fe09ee2961771f0/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/BlobsImpl.java#L107) for reference.

anuchandy commented on sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobServiceImpl.java:
```diff
@@ -1335,6 +1446,96 @@ public void onFailure(Call<ResponseBody> call, Throwable t) {
+        DateTimeRfc1123 ifUnmodifiedSinceConverted = ifUnmodifiedSince == null ? null :
+            new DateTimeRfc1123(ifUnmodifiedSince);
+
+        Call<Void> call = service.delete(containerName,
```
This should be `Call<ResponseBody>` instead of  `Call<Void>`.

anuchandy replied:
The reason for "ResponseBody" is, we want later read the body if the delete fails i.e. if service returns non-202.

anuchandy commented on sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobServiceImpl.java:
```diff
@@ -1485,5 +1686,19 @@ T getResult() {
                                            @Header("x-ms-encryption-algorithm") EncryptionAlgorithmType encryptionAlgorithm);
+
+        @DELETE("{containerName}/{blob}")
+        Call<Void> delete(@Path("containerName") String containerName,
```
return type should be `Call <ResponseBody>`

vcolin7 replied:
I assume we should also do this for APIs like `getBlobProperties` where we don't expect a body in a successful call, but there could be one in a failed call, right?

vcolin7 replied:
Also, the reason I went with `Void` is because the Java API returns a `Mono<Void>` for the [delete operation](https://github.com/Azure/azure-sdk-for-java/blob/712e35f4ad597d43d34c678cc263dbc9095a3ed6/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java#L1071) and the [`BlobDeleteResponse`](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/models/BlobsDeleteResponse.java) object.

vcolin7 replied:
I guess this means we want to return a `Void` at the public API level but a `ResponseBody` at the service interface level. Is this correct?

anuchandy replied:
_I assume we should also do this for APIs like getBlobProperties where we don't expect a body in a successful call, but there could be one in a failed call, right?_

Yes, it seems there is a similar bug in `getBlobProperties` as well, that needs to be fixed.

_I guess this means we want to return a Void at the public API level but a ResponseBody at the service interface level. Is this correct?_

right, like java SDK, from public API we return `Void` but in internal service interface we need `ResponseBody` so that we can read error body from it and throw.

anuchandy commented on sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobServiceImpl.java:
```diff
@@ -1335,6 +1446,96 @@ public void onFailure(Call<ResponseBody> call, Throwable t) {
+            requestId);
+
+        if (callback != null) {
+            executeCall(call, new retrofit2.Callback<Void>() {
```
`new retrofit2.Callback<ResponseBody>()`

anuchandy commented on sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobServiceImpl.java:
```diff
@@ -1335,6 +1446,96 @@ public void onFailure(Call<ResponseBody> call, Throwable t) {
+                            callback.onResponse(new BlobDeleteResponse(response.raw().request(),
+                                response.code(),
+                                response.headers(),
+                                response.body(),
```
since we need to change from `Call<Void>` to `Call<ResponseBody>` here instead of `response.body()` it should be `null`.

vcolin7 commented on sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java:
```diff
@@ -547,6 +551,161 @@ public ServiceCall commitBlockListWithRestResponse(String containerName,
+     * @param blobName      The blob name.
+     * @return The blob's metadata.
+     */
+    Void delete(String containerName,
```
Thank you, I took the template from `getBlobProperties` and forgot to modify it.

vcolin7 commented on sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobClient.java:
```diff
@@ -547,6 +551,161 @@ public ServiceCall commitBlockListWithRestResponse(String containerName,
+    }
+
+    /**
+     * If the storage account's soft delete feature is disabled then, when a blob is deleted, it is permanently
```
I think you are right. I'll add a more brief explanation.

vcolin7 commented:
Thanks for the comments @anuchandy. I have addressed them, made a couple corrections for how other we generate error messages for other operations and added an `undelete` API.

anuchandy commented:
@vcolin7 thank you!. Btw is `undelete` included in the set of preview API we're going to support?

vcolin7 commented:
@anuchandy `undelete` is not in the set of APIs we planned for preview 1, I saw it mentioned in the documentation for the `delete` APi in Java and thought adding it wouldn't hurt, since it's people might want to have the ability to undelete a blob if they can delete one. Any thoughts on this @bsiegel?

anuchandy commented:
@vcolin7 as discussed in sync-up let's remove `undelete` from this PR so that we can get this PR in. 

The `undelete` can be a new PR and exposing it in preview1 can be discussed with Adrian once he is back.
